### PR TITLE
[SPARK-47602][CORE] Resource managers: Migrate logError with variables to structured logging framework

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to:
- migrate `logError` in module `Resource managers` with variables to `structured logging framework`.
- add some test case to `*LoggingSuite`
- add new UT `MDCSuite`
- make the `type` of `value` from `String` to `Any`( so that when using `MDC`, some code can be `omitted`, eg: `String.valueOf(...)` , `x.toString`)

### Why are the changes needed?
To enhance Apache Spark's logging system by implementing structured logging.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass GA.
- Add & update new UT.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/utils/src/main/scala/org/apache/spark/internal/Logging.scala [hunk 1]
```diff
@@ -29,14 +29,15 @@ import org.apache.logging.log4j.core.filter.AbstractFilter
...
+import org.apache.spark.internal.LogKey.LogKey
 import org.apache.spark.util.SparkClassUtils
 
 /**
  * Mapped Diagnostic Context (MDC) that will be used in log messages.
  * The values of the MDC will be inline in the log message, while the key-value pairs will be
  * part of the ThreadContext.
  */
-case class MDC(key: LogKey.Value, value: String)
+case class MDC(key: LogKey, value: Any)
```
Comment (panbingkun) [inline common/utils/src/main/scala/org/apache/spark/internal/Logging.scala:40]: Make the `type` of `value` from `String` to `Any`, so that when using `MDC`, some code can be `omitted`, eg: `String.valueOf(...)` , `x.toString`

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala [hunk 2]
```diff
@@ -57,12 +58,14 @@ trait LoggingSuiteBase
 
   def msgWithMDCAndException: LogEntry = log"Error in executor ${MDC(EXECUTOR_ID, "1")}."
 
-  def expectedPatternForBasicMsg(level: Level): String
-
   def msgWithConcat: LogEntry = log"Min Size: ${MDC(MIN_SIZE, "2")}, " +
     log"Max Size: ${MDC(MAX_SIZE, "4")}. " +
     log"Please double check."
 
+  def expectedPatternForBasicMsg(level: Level): String
```
Comment (panbingkun) [inline common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:65]: Add a new test for `basicMsg` + `Exception`

**Comment by panbingkun** [comment]
cc @gengliangwang

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala [hunk 3]
```diff
@@ -57,12 +58,14 @@ trait LoggingSuiteBase
...
+
+  def expectedPatternForBasicMsgWithException(level: Level): String
```
Comment (panbingkun) [inline common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:67]: Add a new `test` for `basicMsg` + `Exception`

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 4]
```diff
@@ -21,5 +21,9 @@ package org.apache.spark.internal
  * All structured logging keys should be defined here for standardization.
  */
 object LogKey extends Enumeration {
-  val EXECUTOR_ID, MIN_SIZE, MAX_SIZE = Value
+  val APPLICATION_ID, APPLICATION_STATE, BUCKET, CONTAINER_ID, EXECUTOR_ID, POD_ID = Value
```
Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:24]: Each enum can take one line. 
val APPLICATION_ID = Value
val APPLICATION_STATE = Value

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 5]
```diff
@@ -21,5 +21,9 @@ package org.apache.spark.internal
...
+  val MIN_SIZE, MAX_SIZE, MAX_EXECUTOR_FAILURES = Value
+  val EXIT_CODE = Value
+
+  type LogKey = Value
```
Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:28]: why do we need this?

File: common/utils/src/test/scala/org/apache/spark/util/MDCSuite.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.spark.internal.{Logging, MDC}
+import org.apache.spark.internal.LogKey.EXIT_CODE
+
+class MDCSuite
+    extends AnyFunSuite // scalastyle:ignore funsuite
+    with Logging {
+
+  test("check MDC message") {
+    val log = log"This is a log, exitcode ${MDC(EXIT_CODE, 10086)}"
+    assert(log.message === "This is a log, exitcode 10086")
```
Comment (gengliangwang) [inline common/utils/src/test/scala/org/apache/spark/util/MDCSuite.scala:31]: Let's verify the returned map as well

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala [hunk 7]
```diff
@@ -33,7 +33,7 @@ import org.apache.spark.deploy.k8s.Config._
 import org.apache.spark.deploy.k8s.Constants._
 import org.apache.spark.deploy.k8s.KubernetesConf
 import org.apache.spark.deploy.k8s.KubernetesUtils.addOwnerReference
-import org.apache.spark.internal.Logging
+import org.apache.spark.internal.{Logging, LogKey => LK, MDC}
```
Comment (gengliangwang) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:36]: Why alias as LK? LogKey is fine

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala (hunk 3 above)
Comment (gengliangwang) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:67]: Nit: Let's add comments about each pattern is about

**Comment by gengliangwang** [comment]
@panbingkun Thanks for the work. LGTM except some minor comments.

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala (hunk 3 above)
Comment (panbingkun) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:67]: Okay

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala (hunk 4 above)
Comment (panbingkun) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:24]: Yeah

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala (hunk 5 above)
Comment (panbingkun) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:28]: Because I want to use 'xxx: LogKey' instead of 'xxx: LogKey.Value' to define `a type`, it feels like it's a `personal preference`. Should we `restore` it?
https://github.com/apache/spark/blob/c3da26051d0c4c957e1bca98f31fcd575d3a1054/core/src/main/scala/org/apache/spark/scheduler/TaskLocality.scala#L23-L32

File: common/utils/src/test/scala/org/apache/spark/util/MDCSuite.scala (hunk 6 above)
Comment (panbingkun) [reply common/utils/src/test/scala/org/apache/spark/util/MDCSuite.scala:31]: Okay

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala (hunk 7 above)
Comment (panbingkun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:36]: Okay

**Comment by panbingkun** [comment]
I will update `soon` so that we can proceed smoothly with the following work.

**Comment by panbingkun** [comment]
> @panbingkun Thanks for the work. LGTM except some minor comments.

Done.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 8]
```diff
@@ -854,7 +856,8 @@ private[spark] class ApplicationMaster(
             logInfo(s"Driver terminated or disconnected! Shutting down. $remoteAddress")
             finish(FinalApplicationStatus.SUCCEEDED, ApplicationMaster.EXIT_SUCCESS)
           } else {
-            logError(s"Driver terminated with exit code ${exitCode}! Shutting down. $remoteAddress")
+            logError(log"Driver terminated with exit code ${MDC(EXIT_CODE, exitCode)}! " +
+              log"Shutting down. ${MDC(REMOTE_ADDRESS, remoteAddress)}")
```
Comment (panbingkun) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:860]: @gengliangwang 
Perhaps it would be more reasonable to write this in this way
```
logError(log"Driver terminated with exit code ${MDC(EXIT_CODE, exitCode)}! " + log"Shutting down. $remoteAddress")
```
`Not every variable` needs to be `recorded` in the field `content` of `JSON`
Unfortunately, we currently do `not support` this syntax as above, it will `fail` to compile.
So I submitted a separate PR to address this issue: https://github.com/apache/spark/pull/45813

Comment (gengliangwang) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:860]: @panbingkun Let's put all the variables into the context during the migration. If the context is too verbose, we can configure log4j to only show a subset of keys later.
Otherwise, there will be back-and-forth during the migration. `remoteAddress` can be helpful for debugging here.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 8 above)
Comment (panbingkun) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:860]: Okay

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala [hunk 9]
```diff
@@ -532,7 +533,8 @@ class ExecutorPodsAllocator(
       currentTime - creationTime > executorIdleTimeout
     } catch {
       case e: Exception =>
-        logError(s"Cannot get the creationTimestamp of the pod: ${state.pod}", e)
+        logError(log"Cannot get the creationTimestamp of the pod: " +
+          log"${MDC(LogKey.POD_ID, state.pod)}", e)
```
Comment (pan3793) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:537]: IMO `POD_ID` is not a suitable key here. `state.pod` actually outputs a whole Pod Spec, and there is no such a "pod id" concept in the K8s, in practice, we use `namespace` and `pod name` to identify a pod.

Comment (gengliangwang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:537]: @pan3793 Thanks for the suggestion. Do you mind creating a PR to improve this?

Comment (pan3793) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:537]: I would like to have a try, it may take a few hours to learn this new log framework.

Comment (gengliangwang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:537]: Thank you in advance!

Comment (panbingkun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:537]: @pan3793 Thank you very much for `correcting` this issue.
